File: src/com/mishiranu/instantimage/ImageProvider.java
package com.mishiranu.instantimage;

import java.io.FileNotFoundException;
import java.util.ArrayList;

import android.content.ContentProvider;
import android.content.ContentValues;
import android.content.UriMatcher;
import android.database.Cursor;
import android.database.MatrixCursor;
import android.database.sqlite.SQLiteException;
import android.net.Uri;
import android.os.ParcelFileDescriptor;
import android.provider.BaseColumns;
import android.provider.OpenableColumns;

import com.mishiranu.instantimage.util.FileManager;
import com.mishiranu.instantimage.util.FileManager.FileItem;

public class ImageProvider extends ContentProvider
{
	private static final String AUTHORITY = "com.mishiranu.providers.instantimage";
	private static final String PATH_IMAGES = "images";
	
	private static final int URI_IMAGES = 1;
	private static final int URI_IMAGES_ID = 2;
	
	private static final UriMatcher URI_MATCHER;
	
	static
	{
		URI_MATCHER = new UriMatcher(UriMatcher.NO_MATCH);
		URI_MATCHER.addURI(AUTHORITY, PATH_IMAGES, URI_IMAGES);
		URI_MATCHER.addURI(AUTHORITY, PATH_IMAGES + "/#", URI_IMAGES_ID);
	}
	
	public static Uri build(int contentId)
	{
		return Uri.parse("content://" + AUTHORITY + "/" + PATH_IMAGES + "/" + contentId);
	}
	
	@Override
	public boolean onCreate()
	{
		return true;
	}
	
	@Override
	public String getType(Uri uri)
	{
		switch (URI_MATCHER.match(uri))
		{
			case URI_IMAGES: return "vnd.android.cursor.dir/" + AUTHORITY + "." + PATH_IMAGES;
			case URI_IMAGES_ID:
			{
				FileManager.FileItem fileItem = FileManager.findFile(getContext(),
						Integer.parseInt(uri.getLastPathSegment()));
				if (fileItem != null) return fileItem.getMimeType();
				return "application/octet-stream";
			}
			default: throw new IllegalArgumentException("Unknown URI: " + uri);
		}
	}
	
	@Override
	public ParcelFileDescriptor openFile(Uri uri, String mode) throws FileNotFoundException
	{
		switch (URI_MATCHER.match(uri))
		{
			case URI_IMAGES_ID:
			{
				if (!"r".equals(mode)) throw new FileNotFoundException();
				FileManager.FileItem fileItem = FileManager.findFile(getContext(),
						Integer.parseInt(uri.getLastPathSegment()));
				if (fileItem != null)
				{
					return ParcelFileDescriptor.open(fileItem.file, ParcelFileDescriptor.MODE_READ_ONLY);
				}
			}
			default: throw new FileNotFoundException();
		}
	}
	
	private static final String[] PROJECTION = {BaseColumns._ID, OpenableColumns.DISPLAY_NAME, OpenableColumns.SIZE};
	
	@Override
	public Cursor query(Uri uri, String[] projection, String selection, String[] selectionArgs, String sortOrder)
	{
		boolean multiple = false;
		switch (URI_MATCHER.match(uri))
		{
			case URI_IMAGES:
			{
				multiple = true;
			}
			case URI_IMAGES_ID:
			{
				if (projection == null) projection = PROJECTION;
				for (String column : projection)
				{
					switch (column)
					{
						case BaseColumns._ID: break;
						case OpenableColumns.DISPLAY_NAME: break;
						case OpenableColumns.SIZE: break;
						default: throw new SQLiteException("No such column: " + column);
					}
				}
				MatrixCursor cursor = new MatrixCursor(projection);
				int contentId = multiple ? -1 : Integer.parseInt(uri.getLastPathSegment());
				ArrayList<FileManager.FileItem> fileItems = FileManager.listFiles(getContext());
				Object[] values = new Object[projection.length];
				for (FileItem fileItem : fileItems)
				{
					if (contentId == -1 || fileItem.contentId == contentId)
					{
						for (int i = 0; i < projection.length; i++)
						{
							switch (projection[i])
							{
								case BaseColumns._ID:
								{
									values[i] = fileItem.contentId;
									break;
								}
								case OpenableColumns.DISPLAY_NAME:
								{
									values[i] = fileItem.displayName;
									break;
								}
								case OpenableColumns.SIZE:
								{
									values[i] = fileItem.file.length();
									break;
								}
							}
						}
						cursor.addRow(values);
					}
				}
				return cursor;
			}
			default: throw new IllegalArgumentException("Unknown URI: " + uri);
		}
	}
	
	@Override
	public Uri insert(Uri uri, ContentValues values)
	{
		throw new SQLiteException("Unsupported operation");
	}
	
	@Override
	public int update(Uri uri, ContentValues values, String selection, String[] selectionArgs)
	{
		throw new SQLiteException("Unsupported operation");
	}
	
	@Override
	public int delete(Uri uri, String selection, String[] selectionArgs)
	{
		throw new SQLiteException("Unsupported operation");
	}
}

File: src/com/mishiranu/instantimage/ui/ImageProvider.scala
package com.mishiranu.instantimage.ui

import java.io.FileNotFoundException

import android.content.{ContentProvider, ContentValues, UriMatcher}
import android.database.{Cursor, MatrixCursor}
import android.database.sqlite.SQLiteException
import android.net.Uri
import android.os.ParcelFileDescriptor
import android.provider.{BaseColumns, OpenableColumns}
import com.mishiranu.instantimage.util.FileManager

class ImageProvider extends ContentProvider {
  override def onCreate(): Boolean = true

  override def getType(uri: Uri): String = {
    ImageProvider.URI_MATCHER.`match`(uri) match {
      case ImageProvider.URI_IMAGES =>
        "vnd.android.cursor.dir/" + ImageProvider.AUTHORITY + "." + ImageProvider.PATH_IMAGES
      case ImageProvider.URI_IMAGES_ID =>
        FileManager.findFile(getContext, uri.getLastPathSegment.toInt)
          .map(_.mimeType).getOrElse("application/octet-stream")
      case _ =>
        throw new IllegalArgumentException("Unknown URI: " + uri)
    }
  }

  override def openFile(uri: Uri, mode: String): ParcelFileDescriptor = {
    ImageProvider.URI_MATCHER.`match`(uri) match {
      case ImageProvider.URI_IMAGES_ID =>
        if (!"r".equals(mode)) {
          throw new FileNotFoundException
        }
        FileManager.findFile(getContext, Integer.parseInt(uri.getLastPathSegment))
          .map(f => ParcelFileDescriptor.open(f.file, ParcelFileDescriptor.MODE_READ_ONLY))
          .getOrElse(throw new FileNotFoundException)
      case _ =>
        throw new FileNotFoundException
    }
  }

  override def query(uri: Uri, projection: Array[String], selection: String,
    selectionArgs: Array[String], sortOrder: String): Cursor = {
    val matchResult = ImageProvider.URI_MATCHER.`match`(uri)
    val multiple = matchResult == ImageProvider.URI_IMAGES
    if (matchResult == ImageProvider.URI_IMAGES || matchResult == ImageProvider.URI_IMAGES_ID) {
      val workProjection = if (projection != null) projection else ImageProvider.PROJECTION

      workProjection
        .filterNot(_ == BaseColumns._ID)
        .filterNot(_ == OpenableColumns.DISPLAY_NAME)
        .filterNot(_ == OpenableColumns.SIZE)
        .headOption.map(c => throw new SQLiteException("No such column: " + c))

      val cursor = new MatrixCursor(workProjection)
      val contentId = if (multiple) -1 else uri.getLastPathSegment.toInt

      FileManager.listFiles(getContext)
        .filter(f => contentId == -1 || f.contentId == contentId)
        .map{ fileItem =>
        Map(BaseColumns._ID -> fileItem.contentId, OpenableColumns.DISPLAY_NAME -> fileItem.displayName,
          OpenableColumns.SIZE -> fileItem.file.length)
      }.foreach(map => cursor.addRow(workProjection.map(c => map(c).asInstanceOf[AnyRef])))
      cursor
    } else {
      throw new IllegalArgumentException("Unknown URI: " + uri)
    }
  }

  override def insert(uri: Uri, values: ContentValues): Uri = {
    throw new SQLiteException("Unsupported operation")
  }

  override def update(uri: Uri, values: ContentValues, selection: String, selectionArgs: Array[String]): Int = {
    throw new SQLiteException("Unsupported operation")
  }

  override def delete(uri: Uri, selection: String, selectionArgs: Array[String]): Int = {
    throw new SQLiteException("Unsupported operation")
  }
}

object ImageProvider {
  private val AUTHORITY = "com.mishiranu.providers.instantimage"
  private val PATH_IMAGES = "images"

  private val URI_IMAGES = 1
  private val URI_IMAGES_ID = 2

  private val URI_MATCHER = {
    val matcher = new UriMatcher(UriMatcher.NO_MATCH)
    matcher.addURI(AUTHORITY, PATH_IMAGES, URI_IMAGES)
    matcher.addURI(AUTHORITY, PATH_IMAGES + "/#", URI_IMAGES_ID)
    matcher
  }

  def buildUri(contentId: Int): Uri = {
    Uri.parse("content://" + AUTHORITY + "/" + PATH_IMAGES + "/" + contentId)
  }

  private def PROJECTION = Array(BaseColumns._ID, OpenableColumns.DISPLAY_NAME, OpenableColumns.SIZE)
}
